diff -urNp old/agent/mibgroup/hardware/fsys/fsys_getfsstats.c new/agent/mibgroup/hardware/fsys/fsys_getfsstats.c
--- old/agent/mibgroup/hardware/fsys/fsys_getfsstats.c	2018-06-21 16:03:13.675537036 +0200
+++ new/agent/mibgroup/hardware/fsys/fsys_getfsstats.c	2018-06-21 16:10:36.212652288 +0200
@@ -88,6 +88,8 @@ _fs_type( char *typename )
        return NETSNMP_FS_TYPE_EXT2;
     else if ( !strcmp(typename, MOUNT_NTFS) )
        return NETSNMP_FS_TYPE_NTFS;
+    else if ( !strcmp(typename, MOUNT_ZFS) )
+       return NETSNMP_FS_TYPE_OTHER;
 
        /*
         * NetBSD also recognises the following filesystem types:
diff -urNp old/agent/mibgroup/hardware/fsys/fsys_mntent.c new/agent/mibgroup/hardware/fsys/fsys_mntent.c
--- old/agent/mibgroup/hardware/fsys/fsys_mntent.c	2018-06-21 16:03:13.675537036 +0200
+++ new/agent/mibgroup/hardware/fsys/fsys_mntent.c	2018-06-21 16:12:36.982882380 +0200
@@ -142,6 +142,7 @@ _fsys_type( char *typename )
               !strcmp(typename, MNTTYPE_SIMFS) ||
               !strcmp(typename, MNTTYPE_VZFS) ||
               !strcmp(typename, MNTTYPE_BTRFS) ||
+              !strcmp(typename, MNTTYPE_ZFS) ||
               !strcmp(typename, MNTTYPE_ACFS) ||
               !strcmp(typename, MNTTYPE_LOFS))
        return NETSNMP_FS_TYPE_OTHER;
diff -urNp old/agent/mibgroup/hardware/fsys/mnttypes.h new/agent/mibgroup/hardware/fsys/mnttypes.h
--- old/agent/mibgroup/hardware/fsys/mnttypes.h	2018-06-21 16:03:13.675537036 +0200
+++ new/agent/mibgroup/hardware/fsys/mnttypes.h	2018-06-21 16:13:22.793403430 +0200
@@ -151,6 +151,9 @@
 #ifndef MNTTYPE_BTRFS
 #define MNTTYPE_BTRFS    "btrfs"
 #endif
+#ifndef MNTTYPE_ZFS
+#define MNTTYPE_ZFS      "zfs"
+#endif
 #ifndef MNTTYPE_ACFS
 #define MNTTYPE_ACFS  "acfs"
 #endif
diff -urNp old/agent/mibgroup/hardware/fsys/mounts.h new/agent/mibgroup/hardware/fsys/mounts.h
--- old/agent/mibgroup/hardware/fsys/mounts.h	2018-06-21 16:03:13.675537036 +0200
+++ new/agent/mibgroup/hardware/fsys/mounts.h	2018-06-21 16:14:08.014939906 +0200
@@ -56,6 +56,9 @@
 #ifndef MOUNT_ACFS
 #define MOUNT_ACFS   "acfs"
 #endif
+#ifndef MOUNT_ZFS
+#define MOUNT_ZFS    "zfs"
+#endif
 
 /*
 #ifndef MNTTYPE_AFS
