diff -Npru old/agent/mibgroup/agentx/protocol.c new/agent/mibgroup/agentx/protocol.c
--- old/agent/mibgroup/agentx/protocol.c	2012-10-10 00:28:58.000000000 +0200
+++ new/agent/mibgroup/agentx/protocol.c	2017-01-16 10:05:24.419201871 +0100
@@ -1550,7 +1550,7 @@ agentx_parse(netsnmp_session * session,
              size_t len)
 {
     register u_char *bufp = data;
-    u_char          buffer[SNMP_MAX_MSG_SIZE];
+    u_char          buffer[65535];
     oid             oid_buffer[MAX_OID_LEN], end_oid_buf[MAX_OID_LEN];
     size_t          buf_len = sizeof(buffer);
     size_t          oid_buf_len = MAX_OID_LEN;
